Add support for Alter Index Commands

DimCitus opened the pull request:
DESCRIPTION: Adds support for ALTER INDEX (SET|RESET|RENAME TO) commands
DESCRIPTION: Adds support for setting storage parameters on distributed tables

Currently this branch and Pull Request only contains support for ALTER INDEX ... RENAME TO ... on distributed relations from Citus. Before considering a merge, we're going to see about adding other ALTER INDEX sub-commands (well, SET and RESET).

The last commands to support would then be ALTER INDEX ... SET TABLESPACE ..., and that depends on tablespace support in Citus. It's out of scope here.

In passing other bugs where fixed, so before considering a merge we'll need to split and squash the patches into a good series, to be defined. Also WIP commit messages include GitHub issue and PR numbers, this will need a clean up before merge.

Fixes #354
Fixes #1747 
Fixes #1202
See  #1863  (duplicate? sorry)

DimCitus commented:
About the latest patch, I'm using a much simpler way to get to the relation itself or the relation that the index is created on:

```c
	/*
	 * AlterTableStmt applies also to INDEX relations, and we have support for
	 * SET/SET storage parameters in Citus, so we might have to check for
	 * another relation here.
	 */
	if (get_rel_relkind(leftRelationId) == RELKIND_INDEX)
	{
		leftRelationId = IndexGetRelation(leftRelationId, false);
	}
```

I am not sure that the `PlanRenameStmt` implementation should match this one, because of the convoluted/complex logic around the cases when to support, ignore or error out depending on the commands and if the target relation is managed by Citus or not.

It might be that it was easier to get my head around the mechanisms in place in the second run though, but if we choose to simplify `PlanRenameStmt` I'd argue to keep the level of comments I added thanks to the new implementation.

marcocitus commented on src/test/regress/output/multi_alter_table_statements.source:
```diff
@@ -844,6 +844,156 @@ SELECT relname FROM pg_class WHERE relname LIKE 'lineitem_alter%' ORDER BY relna
+ lineitem_alter_220006 | {fillfactor=40}
+ lineitem_alter_220007 | {fillfactor=40}
+ lineitem_alter_220008 | {fillfactor=40}
+ lineitem_alter_220009 | 
```
Wonder what's happening here.

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1371,11 +1383,17 @@ PlanAlterTableStmt(AlterTableStmt *alterTableStatement, const char *alterTableCo
 {
-	Oid relationId = InvalidOid;
+	Oid objRelationId = InvalidOid; /* SQL Object OID */
+	Oid tblRelationId = InvalidOid; /* Relation OID, maybe not the same. */
```
For readability, we prefer full words over acronyms, i.e. objectRelationId, tableRelationId.

DimCitus replied:
this is changed now

marcocitus commented on src/test/regress/output/multi_alter_table_statements.source:
```diff
@@ -844,6 +844,156 @@ SELECT relname FROM pg_class WHERE relname LIKE 'lineitem_alter%' ORDER BY relna
 
 \c - - - :master_port
+-- verify that we can set and reset storage parameters
+ALTER TABLE lineitem_alter SET(fillfactor=40);
```
We should also verify these commands are correctly propagated when creating new shards (or moving one), for example by running them on a local table and then running create_distributed_table.

For the commands to be applied to a new shard, master_get_table_ddl_events needs to return the appropriate command.

DimCitus replied:
This is now tested too.

DimCitus commented:
This patch should address all the points above and in particular implement the needed support for storage options in `master_get_table_ddl_events`.

The test strange result is because of the `\copy` failure test.

Again the plan is the get the PR all-green then squash the branch to a single commit that we can merge, and I think I will have to rebase against current master, too.

marcocitus commented on src/test/regress/output/multi_alter_table_statements.source:
```diff
@@ -1021,3 +1130,67 @@ SELECT relname FROM pg_class WHERE relname LIKE 'test_table_1%';
+
+\c - - - :master_port
+ALTER INDEX idx_lineitem_linenumber RESET(fillfactor);
+ERROR:  relation "idx_lineitem_linenumber" does not exist
```
should this be `hash_dist_pkey`?

DimCitus replied:
oops, too much copy/pasting around those regression test files, good catch!

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -2733,14 +2807,14 @@ OptionsSpecifyOwnedBy(List *optionList, Oid *ownedByTableId)
-
-	if (renameStmt->renameType == OBJECT_TABCONSTRAINT)
+	if (IsAlterTableRenameStmt(renameStmt) &&
+		renameStmt->renameType == OBJECT_TABCONSTRAINT)
```
Is there anything stopping us from renaming constraints?

DimCitus replied:
All I'm seeing at the moment in the code is this error message in src/backend/distributed/relay/relay_event_utility.c: _"unsafe object type in rename statement"_. I had guessed it's been studied and punt off before already, should we see about adding support for it now?

marcocitus replied:
I don't think it's needed in this PR. We might also leave it to another new hire ;).

DimCitus replied:
Or I'll do it next to close the gap. I kind of read the comment as “here be dragons”, if it's only “didn't get there yet” that's a different story... for another PR ok.

marcocitus commented:
Slightly strange message for unsupported ALTER INDEX:
```
ALTER INDEX test_pkey SET TABLESPACE foo;
ERROR:  alter table command is currently unsupported
```

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -2125,12 +2190,21 @@ ErrorIfUnsupportedAlterTableStmt(AlterTableStmt *alterTableStatement)
 								errmsg("alter table command is currently unsupported"),
 								errdetail("Only ADD|DROP COLUMN, SET|DROP NOT NULL, "
 										  "SET|DROP DEFAULT, ADD|DROP CONSTRAINT, "
+										  "SET (), RESET (), "
```
I wonder whether we should start listing what is not supported.

DimCitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -2125,14 +2194,49 @@ ErrorIfUnsupportedAlterTableStmt(AlterTableStmt *alterTableStatement)
+				switch (leftRelationKind)
+				{
+					case RELKIND_RELATION:
+					{
```
Thanks for catching also this, I was mistakenly trying to minimize the footprint of the patch. I guess I was too eager to be done. Hopefully we have reviews!

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -2125,14 +2201,71 @@ ErrorIfUnsupportedAlterTableStmt(AlterTableStmt *alterTableStatement)
+			case AT_ResetRelOptions:    /* RESET (...) */
+			case AT_ReplaceRelOptions:  /* replace entire option list */
+			{
+				/* no special processing needs to be done by Citus here. */
```
I don't think this comment is accurate. Maybe `/* command is supported by Citus */`?

DimCitus replied:
Right. The only thing we do in “here” is erroring out on unsupported commands, after all...

marcocitus replied:
This has been fixed in `ErrorIfUnsupportedAlterTableStmt`, could we update it here as well? I'm still not sure what is meant by "special processing" :).

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1385,32 +1420,71 @@ PlanRenameStmt(RenameStmt *renameStmt, const char *renameCommand)
+		case OBJECT_COLUMN:
+		case OBJECT_TABCONSTRAINT:
+		{
+			/* the target object is our tblRelationId. */
```
tbl -> table

marcocitus replied:
bump

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1385,32 +1420,71 @@ PlanRenameStmt(RenameStmt *renameStmt, const char *renameCommand)
+			/*
+			 * Nodes that are not supported by Citus: we pass-through to the
+			 * main PostgreSQL executor. Any Citus-supported RenameStmt
+			 * renameType must appear above in the switch, explicitely.
```
explicitely -> explicitly

DimCitus replied:
ok

marcocitus commented on src/backend/distributed/utils/citus_ruleutils.c:
```diff
@@ -972,16 +992,23 @@ generate_qualified_relation_name(Oid relid)
+AppendOptionListToString(StringInfo stringBuffer, List *optionList,
+						 const char *prefix,
+						 const char *separator,
+						 const char *operator)
```
I'm not a huge fan of adding all these parameters. This function is doing something fairly trivial, but now you need to consult the call sites to know what exactly it does. What if we just have separate functions for OPTIONS and WITH since they clearly follow a different syntax?

DimCitus replied:
The alternative is to have two function with almost the same content, but it's only about 30 lines each. Let's try that and see if we like it better (in a patch revision coming soon).

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1385,32 +1420,71 @@ PlanRenameStmt(RenameStmt *renameStmt, const char *renameCommand)
+	}
+
+	isDistributedRelation = IsDistributedTable(tableRelationId);
+
```
super minor style/consistency thingy: we usually don't have an empty line between the `if` statement and the result it checks.

DimCitus replied:
ok

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -2125,14 +2201,71 @@ ErrorIfUnsupportedAlterTableStmt(AlterTableStmt *alterTableStatement)
+						 errdetail("Only ADD|DROP COLUMN, SET|DROP NOT NULL, "
+								   "SET|DROP DEFAULT, ADD|DROP CONSTRAINT, "
+								   "SET (), RESET (), "
+								   "ATTACH|DETACH PARTITION and TYPE subcommands "
```
I suppose RENAME TO is missing from this list, but not sure how long we should allow this list to grow.

DimCitus commented:
Ok I rebased against master's branch and squashed commits together to have a PR that looks like it could be considered for merging. It nows contains 3 patches which I think should be kept separate.

marcocitus commented on src/backend/distributed/utils/citus_ruleutils.c:
```diff
@@ -1004,6 +1022,42 @@ AppendOptionListToString(StringInfo stringBuffer, List *optionList)
+static void
+AppendStorageParametersToString(StringInfo stringBuffer, List *optionList)
+{
+	if (optionList != NIL)
```
We have the same `if` statement at the call site. What's the appropriate place for it?

DimCitus commented:
I now see that if I open the commits in GH there are some comments not reported in the main “discussion” page of the Pull Request. I'm going to go through all of them and then push again.

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -2125,14 +2200,71 @@ ErrorIfUnsupportedAlterTableStmt(AlterTableStmt *alterTableStatement)
+	ListCell *commandCell = NULL;
+
+	/* error out if any of the subcommands are unsupported */
+	foreach(commandCell, commandList)
```
I suppose there can only be one subcommand? Should we make that more explicit by just taking the first item? and asserting that there is 1?
